# Decision Tree Classification Model

# Bank Customer Churn Prediction: An End-to-End Machine Learning Experimentation Pipeline

## Project Overview

Customer churn prediction is one of the most important real-world applications of Machine Learning in the banking and financial sector. Retaining existing customers is significantly more cost-effective than acquiring new ones, making churn analysis a critical business problem for modern organizations.

This project focuses on building a complete end-to-end Machine Learning experimentation pipeline for predicting customer churn using banking customer data. Rather than training a single model and reporting accuracy, this project emphasizes analytical thinking, preprocessing decisions, model evaluation, experimentation, and understanding the impact of different Machine Learning techniques on predictive performance.

---

## Dataset

The dataset used in this project is the Bank Customer Churn Prediction dataset, which contains customer demographic information, banking activity, and churn status used for classification analysis.

Dataset Source:  
[Kaggle - Bank Customer Churn Prediction Dataset](https://www.kaggle.com/datasets/shantanudhakadd/bank-customer-churn-prediction)

---

## The project includes:

- Data preprocessing and feature engineering
- Outlier analysis and business-driven preprocessing decisions
- Categorical feature encoding
- Splitting the dataset into the Train set and Test set
- Feature scaling experiments
- Class imbalance handling techniques
- Model training and evaluation
- Cross-validation for performance stability analysis
- Comparative experimentation across multiple Machine Learning models

A major focus of this project is understanding how preprocessing choices affect model behavior and performance in imbalanced classification problems. Different strategies such as Standard Scaling, Robust Scaling, class weighting, and imbalance handling techniques were explored and critically evaluated rather than applied blindly.

The experiments demonstrate the importance of looking beyond accuracy when evaluating classification models. Metrics such as Precision, Recall, F1-Score, ROC-AUC, and Confusion Matrix analysis were used to assess model effectiveness, particularly for detecting churned customers.

This project was designed not only to build predictive models, but also to develop a deeper understanding of practical Machine Learning workflows used in real-world industry environments.

---

## Project Objectives

- Predict whether a customer is likely to churn
- Analyze the impact of preprocessing techniques on model performance
- Evaluate multiple classification algorithms
- Investigate class imbalance handling methods
- Compare model performance using multiple evaluation metrics
- Build a reproducible and professional Machine Learning experimentation workflow

---

## Technologies and Libraries

- Python
- NumPy
- Pandas
- Scikit-learn
- Imbalanced-learn

---

## Machine Learning Concepts Explored

- Classification
- Logistic Regression
- K-Nearest Neighbors (KNN)
- Support Vector Machine SVM
- Naive Bayes
- Decision Trees
- Random Forest
- One-Hot Encoding
- Feature Scaling
- StandardScaler vs RobustScaler
- Outlier Analysis
- Class Imbalance Handling
- Cross Validation
- Model Evaluation Metrics
- ROC-AUC Analysis


In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# **Data preprocessing**

## Data Cleaning Check Template
This template is designed to quickly assess the quality of any dataset before building machine learning models or performing analysis.

It provides a structured overview of the dataset by checking for common data issues such as:

- Missing values

- Duplicate rows

- Incorrect data types

- Outliers

- Distribution of numerical features

- Categorical feature consistency

**What This Template Does**

- Displays basic dataset information (shape, data types)

- Identifies missing values and duplicates

- Summarizes numerical and categorical features

- Detects potential outliers using the IQR method

- Highlights columns with low unique values for quick inspection

How to Use

1. Load your dataset using Pandas  

2. Call the function:

In [2]:
def data_quality_report(df):

    print("DATA QUALITY REPORT")
    
    # Print a separator line for better readability
    
    print("=" * 50)
    print("BASIC INFO")
    print("=" * 50)
    
    # Show general information about the dataset (columns, data types, non-null values)
    print(df.info())
    
    # Show number of rows and columns
    print("\n" + "=" * 50)
    print("SHAPE OF DATA")
    print("=" * 50)
    print(f"Rows: {df.shape[0]}, Columns: {df.shape[1]}")
    
    # Check for missing (null) values in each column
    print("\n" + "=" * 50)
    print("MISSING VALUES")
    print("=" * 50)
    missing = df.isnull().sum()
    
    # Display only columns that have missing values
    print(missing[missing > 0])
    
    # Check for duplicate rows
    print("\n" + "=" * 50)
    print("DUPLICATES")
    print("=" * 50)
    print(f"Duplicate rows: {df.duplicated().sum()}")
    
    # Display data types of each column
    print("\n" + "=" * 50)
    print("DATA TYPES")
    print("=" * 50)
    print(df.dtypes)
    
    # Summary statistics for numerical columns (mean, std, min, max, etc.)
    print("\n" + "=" * 50)
    print("NUMERICAL SUMMARY")
    print("=" * 50)
    print(df.describe())
    
    # Summary for categorical (object) columns
    print("\n" + "=" * 50)
    print("CATEGORICAL SUMMARY")
    print("=" * 50)
    print(df.describe(include=['object']))
    
    # Show unique values for columns with low number of distinct values
    # Useful for detecting categories, errors, or inconsistencies
    print("\n" + "=" * 50)
    print("UNIQUE VALUES (LOW CARDINALITY)")
    print("=" * 50)
    for col in df.columns:
        if df[col].nunique() < 10:  # Only show columns with few unique values
            print(f"{col}: {df[col].unique()}")
            
    # correlation
    print("\n" + "=" * 50)
    print("CORRELATION MATRIX")
    print("=" * 50)
    print(df.corr(numeric_only=True))
    
    # Detect outliers using the IQR (Interquartile Range) method
    print("\n" + "=" * 50)
    print("OUTLIERS CHECK (IQR METHOD)")
    print("=" * 50)
    
    # Loop through only numerical columns
    for col in df.select_dtypes(include=np.number).columns:
        Q1 = df[col].quantile(0.25)  # 25th percentile
        Q3 = df[col].quantile(0.75)  # 75th percentile
        IQR = Q3 - Q1  # Interquartile range
        
        # Count rows that fall outside the normal range
        outliers = df[(df[col] < Q1 - 1.5 * IQR) | (df[col] > Q3 + 1.5 * IQR)]
        print(f"{col}: {len(outliers)} outliers")

# **Load dataset**
Apply Data Cleaning Check Template

In [ ]:
dataset = pd.read_csv("Churn_Modelling.csv")
data_quality_report(dataset)

DATA QUALITY REPORT
BASIC INFO
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB
None

SHAPE OF DATA
Rows: 10000, Columns: 14

MISSING VALUES
Se

# **Outlier Analysis**
Outlier detection was performed using the IQR method on the numerical features.

### Features with Detected Outliers

- `CreditScore`
- `Age`
- `NumOfProducts`
- `Exited`

### Decision About Outliers

The detected outliers were intentionally retained instead of being removed because these values represent realistic customer behavior rather than data entry errors.

#### CreditScore
Extreme credit scores are valid in banking datasets and may contain important information about customer financial behavior and churn patterns. Removing these observations could reduce the model’s ability to learn from high-risk or low-risk customers.

#### Age
Older customers were detected as statistical outliers due to the IQR method. However, these customers are valid observations and may have significantly different churn behavior compared to younger customers. Therefore, they were preserved.

#### NumOfProducts
Customers with a high number of banking products were identified as outliers. These cases are meaningful from a business perspective because product usage can strongly influence customer retention and churn.

#### Exited
The `Exited` feature represents the target variable of the classification problem, where:

- `0` = Customer stayed
- `1` = Customer churned

Although the IQR method detected many values as outliers, this result is not meaningful because `Exited` is a binary categorical variable rather than a continuous numerical feature.

Outlier detection methods such as IQR are designed for continuous numerical distributions and should not be applied to classification labels.

Therefore, no outlier handling was performed on the `Exited` column, and all observations were retained.

### Additional Notes
- Identifier columns such as:
  - `RowNumber`
  - `CustomerId`

  were removed from the dataset because they do not provide predictive value.

- `Surname`

The `Surname` column was removed from the dataset because it represents a high cardinality categorical feature with a large number of unique customer names.

Customer surnames are unlikely to have a direct causal relationship with churn behavior and may introduce noise into the model.

Additionally, encoding this feature could create unnecessary dimensionality or artificial relationships between categories without providing meaningful predictive value.

Therefore, the `Surname` feature was excluded from the modeling process.

### Conclusion

The outliers were retained to preserve potentially valuable business information and avoid removing legitimate customer behavior patterns. Tree-based models such as Random Forest are generally robust to outliers, making aggressive outlier removal unnecessary for this project.

# **Select Independent Variable (X) and Dependent Variable (y)**

Note: The concepts of Independent and Dependent variables were discussed earlier.

If you would like to review or gain a clearer understanding, please refer to this section: [Go to Independent variable VS Dependent variable Section](https://github.com/Hazem1695/ml-concept-briefs)

In [4]:
# Separate independent variable(features (X)) by selecting all rows and all columns except the last one
X = dataset.iloc[:, 3:-1].values
# Separate the dependent variable (Target(y)) by selecting the last column
y = dataset.iloc[:, -1].values

In [5]:
print(X[0:10, :])

[[619 'France' 'Female' 42 2 0.0 1 1 1 101348.88]
 [608 'Spain' 'Female' 41 1 83807.86 1 0 1 112542.58]
 [502 'France' 'Female' 42 8 159660.8 3 1 0 113931.57]
 [699 'France' 'Female' 39 1 0.0 2 0 0 93826.63]
 [850 'Spain' 'Female' 43 2 125510.82 1 1 1 79084.1]
 [645 'Spain' 'Male' 44 8 113755.78 2 1 0 149756.71]
 [822 'France' 'Male' 50 7 0.0 2 1 1 10062.8]
 [376 'Germany' 'Female' 29 4 115046.74 4 1 0 119346.88]
 [501 'France' 'Male' 44 4 142051.07 2 0 1 74940.5]
 [684 'France' 'Male' 27 2 134603.88 1 1 1 71725.73]]


# **Encoding categorical data**

## Encoding the Indpendent Variable Using One Hot Encoding

Note: I explained One-Hot Encoding earlier.

If you missed it or want a deeper understanding of how it works, you can find it here: [Go to One-Hot Encoding Section](https://github.com/Hazem1695/ml-concept-briefs)

In [6]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
# One-hot encode the 'color' column (index 2) while keeping other features
# Initialize ColumnTransformer to handle categorical data
# [2] targets the 'color' column (index 2)
ct = ColumnTransformer(transformers = [('encoder', OneHotEncoder(sparse_output=False, drop='first'),[1,2])], remainder = 'passthrough')
# Execute transformation and convert to NumPy ndarray for model training
X = np.array(ct.fit_transform(X))

### Categorical Feature Encoding

Categorical variables were transformed into numerical representations using One-Hot Encoding.

#### Geography
The `Geography` feature was encoded using One-Hot Encoding. Since the feature contains three categories (`France`, `Spain`, and `Germany`), two binary columns were created while one category was automatically dropped to avoid the dummy variable trap and multicollinearity.

Encoding representation:

| Germany | Spain | Geography |
|---|---|---|
| 0 | 0 | France |
| 0 | 1 | Spain |
| 1 | 0 | Germany |

#### Gender
The `Gender` feature was also transformed using One-Hot Encoding. Since it contains only two categories (`Female` and `Male`), a single binary column was generated after dropping one category.

Encoding representation:

| Gender |
|---|
| 0 → Female |
| 1 → Male |

These transformations convert categorical variables into numerical representations suitable for machine learning algorithms.

# **Splitting the dataset into the Train set and Test set**

In [7]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0)

In [8]:
print(X_train[0:10, :])
#france  0.0 0.0
#spain   0.0 1.0
#germany 1.0 0.0
#female  0.0
#male    1.0

[[1.0 0.0 0.0 579 39 5 117833.3 3 0 0 5831.0]
 [0.0 0.0 0.0 750 32 5 0.0 2 1 0 95611.47]
 [0.0 1.0 0.0 729 34 9 53299.96 2 1 1 42855.97]
 [0.0 1.0 1.0 689 38 5 75075.14 1 1 1 8651.92]
 [0.0 0.0 1.0 605 52 7 0.0 2 1 1 173952.5]
 [0.0 0.0 0.0 667 37 9 71786.9 2 1 1 67734.79]
 [0.0 0.0 1.0 673 65 0 0.0 1 1 1 85733.33]
 [0.0 0.0 1.0 724 31 5 0.0 1 1 0 134889.95]
 [0.0 0.0 0.0 731 38 10 123711.73 2 1 0 171340.68]
 [0.0 1.0 0.0 484 39 5 0.0 2 1 1 175224.12]]


In [9]:
print(X_test[0:10, :])

[[1.0 0.0 0.0 597 35 8 131101.04 1 1 1 192852.67]
 [0.0 0.0 0.0 523 40 2 102967.41 1 1 0 128702.1]
 [0.0 1.0 0.0 706 42 8 95386.82 1 1 1 75732.25]
 [0.0 0.0 1.0 788 32 4 112079.58 1 0 0 89368.59]
 [1.0 0.0 1.0 706 38 5 163034.82 2 1 1 135662.17]
 [0.0 1.0 0.0 670 57 3 175575.95 2 1 0 99061.75]
 [0.0 1.0 1.0 590 34 0 65812.35 2 0 1 160346.3]
 [0.0 1.0 0.0 636 29 6 157576.47 2 1 1 101102.39]
 [0.0 0.0 0.0 598 64 9 0.0 1 0 1 13181.37]
 [0.0 0.0 0.0 456 63 1 165350.61 2 0 0 140758.07]]


In [10]:
print(y_train)

[1 0 0 ... 0 0 1]


In [11]:
print(y_test)

[0 1 0 ... 0 0 0]


# **Feature Scaling Using StandardScaler**

Note: I explained Feature Scaling earlier.

If you want to understand when and why to use feature scaling, you can check it here: [Go to  When is Feature Scaling Necessary in Regression Models? Section](https://github.com/Hazem1695/ml-concept-briefs)

In [12]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train[:, [3,4,5,6,7,10]] = sc.fit_transform(X_train[:, [3,4,5,6,7,10]])
X_test[:, [3,4,5,6,7,10]] = sc.transform(X_test[:, [3,4,5,6,7,10]])

In [13]:
print(X_train[0:10, :])

[[1.0 0.0 0.0 -0.7355070560983401 0.0152657115778632 0.00886037125109047
  0.6731600273055683 2.5350339381815266 0 0 -1.640809940586845]
 [0.0 0.0 0.0 1.0244271940236245 -0.652609169953666 0.00886037125109047
  -1.2077241651131092 0.8042415361458916 1 0 -0.07927152135221784]
 [0.0 1.0 0.0 0.808294917692857 -0.46178777523037196 1.3932933792339848
  -0.3569370597655732 0.8042415361458916 1 1 -0.9968401203574622]
 [0.0 1.0 1.0 0.39661439134853776 -0.08014498578378383
  0.00886037125109047 -0.009356266822735035 -0.9265508658897434 1 1
  -1.5917460967732924]
 [0.0 0.0 1.0 -0.4679147139745326 1.2556047772792747 0.7010768752425377
  -1.2077241651131092 0.8042415361458916 1 1 1.2833024120144987]
 [0.0 0.0 0.0 0.17019010185916217 -0.17555568314543085 1.3932933792339848
  -0.0618439645862603 0.8042415361458916 1 1 -0.5641264883702068]
 [0.0 0.0 1.0 0.23194218081081006 2.495943842980686 -1.7216808887275274
  -1.2077241651131092 -0.9265508658897434 1 1 -0.25108054753105535]
 [0.0 0.0 1.0 0.7568348

In [14]:
print(X_test[0:10, :])

[[1.0 0.0 0.0 -0.5502508192433965 -0.3663770778687249 1.0471851272382613
  0.8849429650245407 -0.9265508658897434 1 1 1.6120302699224127]
 [0.0 0.0 0.0 -1.311859792980387 0.11067640893951022 -1.0294643847360803
  0.4358670251810849 -0.9265508658897434 1 0 0.4962689058885611]
 [0.0 1.0 0.0 0.5715786150448734 0.30149780366280426 1.0471851272382613
  0.31486377821942724 -0.9265508658897434 1 1 -0.425027850923865]
 [0.0 0.0 1.0 1.4155236940507279 -0.652609169953666 -0.33724788074463313
  0.5813177297920199 -0.9265508658897434 0 0 -0.18785300869207555]
 [1.0 0.0 1.0 0.5715786150448734 -0.08014498578378383 0.00886037125109047
  1.394677836289079 0.8042415361458916 1 1 0.6173243726906659]
 [0.0 1.0 0.0 0.20106614133498613 1.7326582640875097 -0.6833561327403567
  1.5948624460052427 0.8042415361458916 1 0 -0.01926131225978551]
 [0.0 1.0 1.0 -0.6222949113536523 -0.46178777523037196
  -1.7216808887275274 -0.15721120501672425 0.8042415361458916 0 1
  1.0466517903393149]
 [0.0 1.0 0.0 -0.1488623060

# **Decision Tree Hyperparameter Optimization using GridSearchCV**

if you want to try GridSearchCV remove (""") and run cell

In [ ]:
"""

from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV

dt = DecisionTreeClassifier(class_weight='balanced', random_state=0)

param_grid = {'max_depth': [4,5,6,7,8,10], 'min_samples_split': [2,5,10,20], 'min_samples_leaf': [1,2,3,5], 'criterion': ['gini', 'entropy']}

grid_search = GridSearchCV(estimator=dt, param_grid=param_grid, scoring='roc_auc', cv=5, n_jobs=-1, verbose=2)

grid_search.fit(X_train, y_train)

"""

best params we got from using GridSearchCV

In [23]:
print(grid_search.best_params_)

{'criterion': 'entropy', 'max_depth': 6, 'min_samples_leaf': 3, 'min_samples_split': 20}


# **Building Decision Tree Classification Model**

## Training the Building Decision Tree Classification model on the Training set

In [ ]:
from sklearn.tree import DecisionTreeClassifier
classifier = DecisionTreeClassifier(max_depth=6, min_samples_split=3, min_samples_leaf=20, criterion='entropy', class_weight = 'balanced', random_state=0)
classifier.fit(X_train, y_train)

#{'criterion': 'entropy', 'max_depth': 6, 'min_samples_leaf': 3, 'min_samples_split': 20} from GridSearchCV

DecisionTreeClassifier(class_weight='balanced', criterion='entropy',
                       max_depth=6, min_samples_leaf=20, min_samples_split=3,
                       random_state=0)

# **Handling Class Imbalance using `Class_Weight = 'balanced'`**

### Difference Between Handling Class Imbalance Methods

| Method        | Action                             |
| ------------- | ---------------------------------- |
| Class Weights | changes learning importance        |
| SMOTE         | creates synthetic minority samples |
| Undersampling | removes majority samples           |


Note: I explained handling imbalanced data earlier.

If you want to understand when and why to use techniques such as Class Weights, SMOTE, and Undersampling, you can check:
[Handling Imbalanced Data in Machine Learning](https://github.com/Hazem1695/ml-concept-briefs)


# **Preticting the Test set results**

In [25]:
y_pred = classifier.predict(X_test)

print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1)[0:100,:])

[[1 0]
 [0 1]
 [1 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 1]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 0]
 [0 0]
 [0 0]
 [0 0]
 [1 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [1 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [1 0]
 [0 0]
 [0 0]
 [0 0]
 [1 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [1 0]
 [0 0]
 [0 0]
 [1 0]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [0 1]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [1 0]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [1 0]
 [0 0]
 [1 1]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]]


# **Evaluating the Model Performance**

In [26]:
from sklearn.metrics import confusion_matrix, ConfusionMatrixDisplay, accuracy_score, classification_report, roc_auc_score
from sklearn.model_selection import cross_val_score

print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

print("\nAccuracy Score:")
print(accuracy_score(y_test, y_pred))

print("\nClassification Report:")
print(classification_report(y_test, y_pred))

print("\nROC-AUC Score:")
print(roc_auc_score(y_test, y_pred))


accuracies = cross_val_score(estimator=classifier, X=X_train, y=y_train, cv=10)

print("\nMean Accuracy:")
print(accuracies.mean())

print("\nStandard Deviation:")
print(accuracies.std())

Confusion Matrix:
[[1520  471]
 [ 106  403]]

Accuracy Score:
0.7692

Classification Report:
              precision    recall  f1-score   support

           0       0.93      0.76      0.84      1991
           1       0.46      0.79      0.58       509

    accuracy                           0.77      2500
   macro avg       0.70      0.78      0.71      2500
weighted avg       0.84      0.77      0.79      2500


ROC-AUC Score:
0.7775919930453248

Mean Accuracy:
0.7687999999999999

Standard Deviation:
0.019778327083507918


| ROC-AUC   | Quality   |
| --------- | --------- |
| 0.50      | Random    |
| 0.60–0.70 | Weak      |
| 0.70–0.80 | Decent    |
| 0.80–0.90 | Strong    |
| 0.90+     | Excellent |

# Decision Tree Classifier Model Comparison and Evaluation

This experiment focused on evaluating different configurations of Decision Tree models for an imbalanced classification problem. The comparison investigated the impact of:

* Balanced class weights
* Hyperparameter tuning
* Feature scaling strategies
* Tree depth optimization
* Minority-class detection capability
* Model generalization and stability

The experiments aimed to determine which Decision Tree configuration provides the strongest balance between predictive performance and minority-class detection.

---

# 1. Decision Tree Classifier Using StandardScaler Without Balanced Class Weights

The baseline Decision Tree model achieved an accuracy score of `79.56%` and demonstrated moderate performance on the minority class.

The model achieved:

* Minority precision = `0.50`
* Minority recall = `0.54`
* Minority F1-score = `0.52`
* Receiver Operating Characteristic - Area Under Curve = `0.699`

Compared to previous experiments with other algorithms, this baseline Decision Tree model already demonstrated significantly stronger minority-class detection capability without requiring imbalance-handling techniques such as Synthetic Minority Oversampling Technique.

The model maintained a relatively balanced trade-off between majority-class accuracy and minority-class detection.

This experiment highlights one of the advantages of Decision Trees: they naturally handle some imbalance patterns better than probabilistic and linear models.

---

# 2. Decision Tree Classifier Using StandardScaler With Balanced Class Weights

After introducing balanced class weights, the model performance improved across nearly all evaluation metrics.

The model achieved:

* Accuracy score = `81.40%`
* Minority recall = `0.56`
* Minority F1-score = `0.55`
* Receiver Operating Characteristic - Area Under Curve = `0.719`

This configuration improved:

* Minority-class recall
* Minority-class F1-score
* Receiver Operating Characteristic - Area Under Curve
* Overall accuracy

The results demonstrate that balanced class weights successfully encouraged the model to pay more attention to minority-class samples without significantly damaging majority-class performance.

This experiment confirms the effectiveness of class-weight balancing in Decision Tree classifiers.

---

# 3. Decision Tree Classifier Using RobustScaler With Balanced Class Weights

Using RobustScaler instead of StandardScaler produced nearly identical results.

The model achieved:

* Accuracy score = `81.32%`
* Minority recall = `0.56`
* Minority F1-score = `0.55`
* Receiver Operating Characteristic - Area Under Curve = `0.718`

The minimal difference between StandardScaler and RobustScaler indicates that Decision Trees are relatively insensitive to feature scaling because they rely on feature splits rather than distance-based calculations.

This experiment demonstrates an important machine learning concept:

* scaling is generally less critical for tree-based algorithms compared to distance-based models such as Support Vector Machines or K-Nearest Neighbors.

---

# 4. Tuned Decision Tree Classifier With Maximum Depth = 6 Using RobustScaler

After hyperparameter tuning, the Decision Tree model demonstrated a substantial improvement in minority-class detection capability.

The model achieved:

* Accuracy score = `75.72%`
* Minority recall = `0.80`
* Minority F1-score = `0.57`
* Receiver Operating Characteristic - Area Under Curve = `0.772`

This configuration produced one of the strongest recall scores among all Decision Tree experiments.

Although overall accuracy decreased compared to earlier models, the model became significantly better at identifying minority-class samples.

The improvement in Receiver Operating Characteristic - Area Under Curve indicates stronger class separation capability and improved overall predictive balance.

This experiment demonstrates the importance of controlling tree complexity and tuning hyperparameters in Decision Tree classifiers.

---

# 5. Tuned Decision Tree Classifier With Maximum Depth = 7 Using RobustScaler

Increasing the maximum depth to `7` produced slightly improved overall performance.

The model achieved:

* Accuracy score = `76.12%`
* Minority recall = `0.79`
* Minority F1-score = `0.57`
* Receiver Operating Characteristic - Area Under Curve = `0.773`

Compared to the previous configuration, this model maintained strong minority-class detection while slightly improving overall stability and predictive balance.

The Receiver Operating Characteristic - Area Under Curve score became one of the highest among all Decision Tree experiments.

This experiment demonstrates that moderate increases in tree depth can improve the model’s ability to capture more complex decision boundaries.

---

# 6. Tuned Decision Tree Classifier With Maximum Depth = 7 Using StandardScaler

Using StandardScaler instead of RobustScaler produced identical results.

The model achieved:

* Accuracy score = `76.12%`
* Minority recall = `0.79`
* Minority F1-score = `0.57`
* Receiver Operating Characteristic - Area Under Curve = `0.773`

This result further confirms that feature scaling has minimal influence on Decision Tree performance.

The identical metrics strengthen the conclusion that tree-based models are largely independent of scaling techniques.

---

# 7. Best Hyperparameter-Tuned Decision Tree Classifier From Grid Search

The final optimized Decision Tree model, selected using grid search, produced the strongest overall Decision Tree performance.

The model achieved:

* Accuracy score = `76.92%`
* Minority recall = `0.79`
* Minority F1-score = `0.58`
* Receiver Operating Characteristic - Area Under Curve = `0.778`

This configuration provided:

* The highest Receiver Operating Characteristic - Area Under Curve score
* The best minority-class F1-score
* Strong recall performance
* Improved overall predictive balance

The hyperparameter optimization successfully controlled overfitting while maintaining strong sensitivity toward minority-class samples.

This experiment demonstrates the effectiveness of systematic hyperparameter tuning in improving Decision Tree performance for imbalanced classification problems.

---

# Performance Comparison Table

| Experiment                                                    | Accuracy | Minority Precision | Minority Recall | Minority F1-Score | Receiver Operating Characteristic - Area Under Curve |
| ------------------------------------------------------------- | -------- | ------------------ | --------------- | ----------------- | ---------------------------------------------------- |
| Decision Tree Without Balanced Class Weights + StandardScaler | 0.796    | 0.50               | 0.54            | 0.52              | 0.699                                                |
| Decision Tree With Balanced Class Weights + StandardScaler    | 0.814    | 0.54               | 0.56            | 0.55              | 0.719                                                |
| Decision Tree With Balanced Class Weights + RobustScaler      | 0.813    | 0.54               | 0.56            | 0.55              | 0.718                                                |
| Tuned Decision Tree Maximum Depth 6 + RobustScaler            | 0.757    | 0.45               | 0.80            | 0.57              | 0.772                                                |
| Tuned Decision Tree Maximum Depth 7 + RobustScaler            | 0.761    | 0.45               | 0.79            | 0.57              | 0.773                                                |
| Tuned Decision Tree Maximum Depth 7 + StandardScaler          | 0.761    | 0.45               | 0.79            | 0.57              | 0.773                                                |
| Best Grid Search Tuned Decision Tree + StandardScaler         | 0.769    | 0.46               | 0.79            | 0.58              | 0.778                                                |

---

# Ranking of Experiments

| Rank | Experiment                                                    | Main Strength                                                                        |
| ---- | ------------------------------------------------------------- | ------------------------------------------------------------------------------------ |
| 1    | Best Grid Search Tuned Decision Tree + StandardScaler         | Best Receiver Operating Characteristic - Area Under Curve and best minority F1-score |
| 2    | Tuned Decision Tree Maximum Depth 7 + RobustScaler            | Strong minority recall and balanced predictive capability                            |
| 3    | Tuned Decision Tree Maximum Depth 7 + StandardScaler          | Strong recall with stable performance                                                |
| 4    | Tuned Decision Tree Maximum Depth 6 + RobustScaler            | Highest minority recall                                                              |
| 5    | Decision Tree With Balanced Class Weights + StandardScaler    | Strong overall accuracy and balanced performance                                     |
| 6    | Decision Tree With Balanced Class Weights + RobustScaler      | Similar balanced performance                                                         |
| 7    | Decision Tree Without Balanced Class Weights + StandardScaler | Weakest minority detection among Decision Tree experiments                           |

---

# Key Insights

## Decision Trees Naturally Handle Imbalanced Data Better Than Some Other Algorithms

Even without imbalance handling, the baseline Decision Tree model achieved moderate minority-class recall and F1-score performance.

This contrasts strongly with algorithms such as Gaussian Naive Bayes and linear Support Vector Classification models, which completely failed to classify minority-class samples without balancing strategies.

---

## Balanced Class Weights Improved Overall Predictive Balance

Applying balanced class weights improved:

* Minority recall
* F1-score
* Receiver Operating Characteristic - Area Under Curve
* Overall accuracy

This demonstrates that class weighting is highly effective for Decision Tree classifiers.

---

## Hyperparameter Tuning Significantly Improved Minority-Class Detection

Reducing tree complexity through hyperparameter tuning dramatically improved minority recall and Receiver Operating Characteristic - Area Under Curve performance.

This confirms that controlling overfitting is critical for reliable Decision Tree performance.

---

## Feature Scaling Had Minimal Influence on Decision Tree Performance

StandardScaler and RobustScaler produced nearly identical results across multiple experiments.

This demonstrates that Decision Trees are largely insensitive to scaling because they rely on feature thresholds instead of distance calculations.

---

## Receiver Operating Characteristic - Area Under Curve Improved After Tree Optimization

The tuned Decision Tree models achieved substantially stronger Receiver Operating Characteristic - Area Under Curve scores than the baseline models.

This indicates stronger class separation capability and improved generalization performance.

---

# Conclusion

This experimentation demonstrated that Decision Tree performance in imbalanced classification problems is strongly influenced by:

* Class weighting
* Hyperparameter tuning
* Tree complexity control

The experiments showed that:

* Decision Trees naturally handle imbalance better than several other algorithms
* Balanced class weights improve predictive balance and minority-class detection
* Hyperparameter tuning significantly improves Receiver Operating Characteristic - Area Under Curve and recall performance
* Feature scaling has minimal impact on Decision Tree classifiers

The strongest overall configuration was the hyperparameter-tuned Decision Tree model selected through grid search using balanced class weights and StandardScaler, as it achieved the best balance between recall, F1-score, Receiver Operating Characteristic - Area Under Curve, and overall predictive capability.

These experiments highlight the importance of systematic tuning and metric-driven evaluation when building reliable tree-based machine learning classification pipelines for imbalanced datasets.
